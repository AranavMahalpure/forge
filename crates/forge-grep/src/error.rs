use std::error::Error;
use std::fmt;

#[derive(Debug)]
pub enum GrepError {
    IoError(std::io::Error),
    PatternError(String),
}

impl fmt::Display for GrepError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            GrepError::IoError(err) => write!(f, "IO error: {}", err),
            GrepError::PatternError(msg) => write!(f, "Pattern error: {}", msg),
        }
    }
}

impl Error for GrepError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            GrepError::IoError(err) => Some(err),
            GrepError::PatternError(_) => None,
        }
    }
}

impl From<std::io::Error> for GrepError {
    fn from(err: std::io::Error) -> Self {
        GrepError::IoError(err)
    }
}

pub type Result<T> = std::result::Result<T, GrepError>;
