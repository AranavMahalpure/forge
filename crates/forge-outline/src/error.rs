use std::fmt::{Debug, Display, Formatter};
use std::io;

#[derive(Debug)]
pub enum Error {
    DirectoryAccess(String),
    DirectoryRead(io::Error),
    FileRead { path: String, error: io::Error },
    LanguageError(String),
    QueryError(String),
}

pub type Result<T> = std::result::Result<T, Error>;

impl std::error::Error for Error {}

impl Display for Error {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Error::DirectoryAccess(msg) => write!(f, "{}", msg),
            Error::DirectoryRead(e) => write!(f, "Failed to read directory: {}", e),
            Error::FileRead { path, error } => {
                write!(f, "Failed to read file {}: {}", path, error)
            }
            Error::LanguageError(msg) => write!(f, "Language error: {}", msg),
            Error::QueryError(msg) => write!(f, "Query error: {}", msg),
        }
    }
}
