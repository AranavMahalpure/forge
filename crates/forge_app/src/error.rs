use std::fmt::Debug;

use derive_more::derive::Display;
use serde::{Deserialize, Serialize};

#[derive(Debug, Display, derive_more::From)]
pub enum Error {
    Diesel(diesel::result::Error),
    DieselConnection(diesel::ConnectionError),
    DieselR2D2(diesel::r2d2::Error),
    Domain(forge_domain::Error),
    EmptyResponse,
    Handlebars(handlebars::RenderError),
    IO(std::io::Error),
    Provider(forge_provider::Error),
    R2D2(r2d2::Error),
    Serde(serde_json::Error),
    StdError(Box<dyn std::error::Error + Send + Sync>),
    ToolCallMissingName,
    Var(std::env::VarError),
    Walk(forge_walker::Error),
}

impl Error {
    pub fn from_std_error<T: std::error::Error + Send + Sync + 'static>(err: T) -> Self {
        Error::StdError(Box::new(err))
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ErrorResponse {
    pub error: InnerErrorResponse,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct InnerErrorResponse {
    pub code: u32,
    pub message: String,
    pub metadata: Option<std::collections::HashMap<String, serde_json::Value>>,
}

pub type Result<T> = std::result::Result<T, Error>;
