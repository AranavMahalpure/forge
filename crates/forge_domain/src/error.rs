use thiserror::Error;

// NOTE: Deriving From for error is a really bad idea. This is because you end
// up converting errors incorrectly without much context. For eg: You don't want
// all serde error to be treated as the same. Instead we want to know exactly
// where that serde failure happened and for what kind of value.
#[derive(Debug, Error)]
pub enum Error {
    #[error("Missing tool name")]
    ToolCallMissingName,

    #[error("Invalid tool call arguments: {0}")]
    ToolCallArgument(serde_json::Error),

    #[error("Invalid conversation id: {0}")]
    ConversationId(uuid::Error),

    #[error("Invalid input command: {0}")]
    InputCommand(String),

    #[error("Invalid template rendering params: {0}")]
    Template(handlebars::RenderError),
}

pub type Result<A> = std::result::Result<A, Error>;
