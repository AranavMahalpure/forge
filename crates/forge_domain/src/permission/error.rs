//! Error types for the permission system.

use std::path::PathBuf;
use thiserror::Error;

/// Errors that can occur in the permission system.
#[derive(Debug, Error)]
pub enum PermissionError {
    /// Access denied by user
    #[error("User denied access to {0}")]
    UserDenied(PathBuf),

    /// Access denied by system
    #[error("System denied access to {0}")]
    SystemDenied(PathBuf),

    /// Invalid response from user interaction
    #[error("Invalid permission response")]
    InvalidResponse,

    /// Interaction with user failed
    #[error("Permission interaction failed: {0}")]
    InteractionFailed(String),

    /// Timeout during interaction
    #[error("Permission request timed out")]
    InteractionTimeout,

    /// The path is outside the allowed directory
    #[error("Path {0} is outside allowed directory")]
    OutsideAllowedDirectory(PathBuf),

    /// Session has expired
    #[error("Permission session has expired")]
    SessionExpired,

    /// Configuration error
    #[error("Permission config error: {0}")]
    ConfigError(String),

    /// Storage error
    #[error("Permission storage error: {0}")]
    StorageError(String),

    /// IO error
    #[error("IO error: {0}")]
    IoError(#[from] std::io::Error),

    /// Error from file system walker
    #[error("File system error: {0}")]
    WalkerError(#[from] forge_walker::Error),
}

/// Result type for permission operations
pub type PermissionResult<T> = std::result::Result<T, PermissionError>;
