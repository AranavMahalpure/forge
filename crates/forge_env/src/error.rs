use derive_more::derive::{Display, From};

#[derive(Debug, Display, From)]
pub enum Error {
    IO(std::io::Error),
    Var(std::env::VarError),
}

pub type Result<T> = std::result::Result<T, Error>;
