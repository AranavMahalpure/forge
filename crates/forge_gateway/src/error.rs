use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::Json;
use serde_json::json;
use thiserror::Error;

/// Custom error type for the application
#[derive(Debug, Error)]
pub enum Error {
    /// Authentication or authorization error
    #[error("Authentication error: {0}")]
    Auth(String),

    /// Database operation error
    #[error("Database error: {0}")]
    Database(String),

    /// Service error
    #[error("Service error: {0}")]
    Service(String),

    /// Bad requests
    #[error("{0}")]
    BadRequest(#[from] validator::ValidationErrors),

    #[error("Serde error: {0}")]
    SerdeError(#[from] serde_json::Error),

    /// Not found error
    #[error("Not found: {0}")]
    NotFound(String),

    #[error("Internal error: {0}")]
    Internal(String),
}

/// Result type alias using our custom Error
pub type Result<T> = std::result::Result<T, Error>;

impl IntoResponse for Error {
    fn into_response(self) -> Response {
        let (status, message) = match self {
            Error::Auth(msg) => (StatusCode::UNAUTHORIZED, msg),
            Error::Database(msg) => (StatusCode::INTERNAL_SERVER_ERROR, msg),
            Error::Service(msg) => (StatusCode::SERVICE_UNAVAILABLE, msg),
            Error::BadRequest(msg) => (StatusCode::BAD_REQUEST, msg.to_string()),
            Error::NotFound(msg) => (StatusCode::NOT_FOUND, msg),
            Error::SerdeError(err) => (StatusCode::INTERNAL_SERVER_ERROR, err.to_string()),
            Error::Internal(msg) => (StatusCode::INTERNAL_SERVER_ERROR, msg),
        };

        let body = Json(json!({
            "error": {
                "message": message,
            }
        }));

        (status, body).into_response()
    }
}

impl From<forge_domain::Error> for Error {
    fn from(err: forge_domain::Error) -> Self {
        Error::Service(err.to_string())
    }
}
