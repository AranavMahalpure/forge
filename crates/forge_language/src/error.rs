#[derive(Debug, thiserror::Error)]
pub enum LanguageError {
    #[error("Parser initialization failed: {0}")]
    ParserInitError(String),
    
    #[error("Failed to parse source: {0}")]
    ParseError(String),
    
    #[error("Invalid query pattern: {0}")]
    QueryError(String),
    
    #[error("Language not supported: {0}")]
    UnsupportedLanguage(String),
}

pub type Result<T> = std::result::Result<T, LanguageError>;