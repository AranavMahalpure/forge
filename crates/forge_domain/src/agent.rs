use std::fmt;
use std::marker::PhantomData;
use std::path::PathBuf;

use async_trait::async_trait;
use derive_setters::Setters;
use handlebars::Handlebars;
use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};

use crate::error::Result;

/// Represents the contents of a prompt, which can either be direct text or a
/// file reference
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum PromptContent {
    Text(String),
    File(PathBuf),
}

impl From<String> for PromptContent {
    fn from(s: String) -> Self {
        PromptContent::Text(s)
    }
}

impl From<&str> for PromptContent {
    fn from(s: &str) -> Self {
        PromptContent::Text(s.to_string())
    }
}

impl From<PathBuf> for PromptContent {
    fn from(p: PathBuf) -> Self {
        PromptContent::File(p)
    }
}

impl fmt::Display for PromptContent {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            PromptContent::Text(content) => write!(f, "{}", content),
            PromptContent::File(path) => write!(f, "@{}", path.display()),
        }
    }
}

impl PromptContent {
    pub fn new(content: impl Into<String>) -> Self {
        PromptContent::Text(content.into())
    }

    pub fn from_file<P: AsRef<std::path::Path>>(path: P) -> Self {
        PromptContent::File(path.as_ref().to_path_buf())
    }
}

/// Represents an AI agent with specific system and user prompts, templated with
/// a context type
#[derive(Clone, Debug, Default, Deserialize, Serialize, Setters)]
#[serde(rename_all = "camelCase")]
#[setters(into, strip_option)]
pub struct Agent<C> {
    /// Name of the agent
    pub name: String,

    /// Description of what the agent does
    #[serde(rename = "description", skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,

    /// The system prompt that defines the agent's behavior
    #[serde(rename = "systemPrompt", skip_serializing_if = "Option::is_none")]
    pub system_prompt: Option<PromptContent>,

    /// Optional user prompt template for consistent interactions
    #[serde(rename = "userPrompt", skip_serializing_if = "Option::is_none")]
    pub user_prompt: Option<PromptContent>,

    /// Phantom data for the context type
    #[serde(skip)]
    _context: PhantomData<C>,
}

impl<C> Agent<C>
where
    C: Serialize + DeserializeOwned,
{
    /// Creates a new agent
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            description: None,
            system_prompt: None,
            user_prompt: None,
            _context: PhantomData,
        }
    }

    /// Renders the system prompt with the given context if it exists
    pub fn render_system_prompt(&self, context: &C) -> Result<Option<String>> {
        if let Some(system_prompt) = &self.system_prompt {
            let handlebars = Handlebars::new();
            let prompt = system_prompt.to_string();
            Ok(Some(handlebars.render_template(&prompt, context)?))
        } else {
            Ok(None)
        }
    }

    /// Renders the user prompt with the given context if it exists
    pub fn render_user_prompt(&self, context: &C) -> Result<Option<String>> {
        if let Some(user_prompt) = &self.user_prompt {
            let handlebars = Handlebars::new();
            let prompt = user_prompt.to_string();
            Ok(Some(handlebars.render_template(&prompt, context)?))
        } else {
            Ok(None)
        }
    }
}

/// Loader trait for loading Agent definitions. The loader should be able to
/// load all available agents from a source (e.g., configuration files).
#[async_trait]
pub trait AgentLoader<C>
where
    C: Send + Sync + DeserializeOwned,
{
    /// Load all available agents. Returns a Vec of agents in the order they
    /// were defined in the source.
    async fn load_agents(&self) -> anyhow::Result<Vec<Agent<C>>>;
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;
    use serde::Deserialize;

    use super::*;

    #[derive(Debug, Serialize, Deserialize)]
    struct CodeContext {
        role: String,
        language: String,
    }

    use std::path::Path;

    #[test]
    fn test_create_agent() {
        let agent: Agent<CodeContext> = Agent::new("Coder");

        assert_eq!(agent.name, "Coder");
        assert_eq!(agent.description, None);
        assert_eq!(agent.system_prompt, None);
        assert_eq!(agent.user_prompt, None);
    }

    #[test]
    fn test_create_with_prompts() {
        let agent: Agent<CodeContext> = Agent::new("Coder")
            .description("A coding assistant")
            .system_prompt("You are a helpful coding assistant");

        assert_eq!(agent.name, "Coder");
        assert_eq!(agent.description, Some("A coding assistant".to_string()));
        assert_eq!(
            agent.system_prompt,
            Some(PromptContent::Text(
                "You are a helpful coding assistant".to_string()
            ))
        );
        assert_eq!(agent.user_prompt, None);
    }

    #[test]
    fn test_agent_with_file_prompts() {
        let agent: Agent<CodeContext> = Agent::new("Coder")
            .description("A coding assistant")
            .system_prompt(PromptContent::from_file("prompts/system.md"))
            .user_prompt(PromptContent::from_file("prompts/user.md"));

        assert_eq!(
            agent.system_prompt,
            Some(PromptContent::File(
                Path::new("prompts/system.md").to_path_buf()
            ))
        );
        assert_eq!(
            agent.user_prompt,
            Some(PromptContent::File(
                Path::new("prompts/user.md").to_path_buf()
            ))
        );
    }

    #[test]
    fn test_render_prompts_with_context() {
        let agent: Agent<CodeContext> = Agent::new("Coder")
            .description("A coding assistant")
            .system_prompt("You are a {{role}} coding assistant")
            .user_prompt("How can I help with {{language}} code today?");

        let context = CodeContext { role: "helpful".to_string(), language: "Rust".to_string() };

        assert_eq!(
            agent.render_system_prompt(&context).unwrap(),
            Some("You are a helpful coding assistant".to_string())
        );
        assert_eq!(
            agent.render_user_prompt(&context).unwrap(),
            Some("How can I help with Rust code today?".to_string())
        );
    }

    #[test]
    fn test_agent_json_format() {
        // Test full agent with all fields
        let full_agent: Agent<CodeContext> = Agent::new("Coder")
            .description("A coding assistant")
            .system_prompt(PromptContent::Text(
                "You are a helpful coding assistant".to_string(),
            ))
            .user_prompt("How can I help?");

        insta::assert_json_snapshot!(full_agent);

        // Test minimal agent (only name)
        let minimal_agent: Agent<CodeContext> = Agent::new("Coder");

        insta::assert_json_snapshot!(minimal_agent);

        // Test agent with file prompts
        let file_agent: Agent<CodeContext> = Agent::new("Coder")
            .description("A coding assistant")
            .system_prompt(PromptContent::from_file("prompts/system.md"))
            .user_prompt(PromptContent::from_file("prompts/user.md"));

        insta::assert_json_snapshot!(file_agent);
    }

    #[test]
    fn test_agent_roundtrip() {
        let agent: Agent<CodeContext> = Agent::new("Coder")
            .description("A coding assistant")
            .system_prompt(PromptContent::Text("System prompt".to_string()))
            .user_prompt("User prompt");

        let serialized = serde_json::to_string(&agent).unwrap();
        let deserialized: Agent<CodeContext> = serde_json::from_str(&serialized).unwrap();

        assert_eq!(deserialized.name, agent.name);
        assert_eq!(deserialized.description, agent.description);
        assert_eq!(deserialized.system_prompt, agent.system_prompt);
        assert_eq!(deserialized.user_prompt, agent.user_prompt);
    }
}
