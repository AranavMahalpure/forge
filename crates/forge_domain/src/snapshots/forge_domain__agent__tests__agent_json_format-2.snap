---
source: crates/forge_domain/src/agent.rs
expression: minimal_agent
snapshot_kind: text
---
{
  "name": "Coder"
}
